internal/bzlmod/go_deps.bzl

```diff
@@ -109,11 +121,37 @@ def _go_deps_impl(module_ctx):
                 ),
             )
 
+    # All `replace` directives are applied after version resolution.
+    # We can simply do this by checking the replace paths' existence
+    # in the module resolutions and swapping out the entry.
+    for path, replace in replace_map.items():
+        if path in module_resolutions:
+            module = module_resolutions[path]
+            module_resolutions[path] = struct(
+                module = module.module,
+                repo_name = module.repo_name,
+                replace = replace.to_path,
+                version = semver.to_comparable(replace.version),
+                raw_version = replace.version,
+                build_naming_convention = module.build_naming_convention,
+                build_file_proto_mode = module.build_file_proto_mode,
+            )
+
+
+    # Validate all the modules have valid sums after the version resolution.
+    # Modules that are discarded in resolution do not need to have sums defined
+    # in the root module.
+    for path, module in module_resolutions.items():
+        entry = (module.replace, module.raw_version) if hasattr(module, "replace") else (path, module.raw_version)
```

Comment:
Could use a helper function that gets the sum (or None) taking replacement into account. That function could then be used below.